### PR TITLE
feat: module `avm/res/managed-services/registration-definition`

### LINKED ISSUE #558: [Module Proposal]: `avm/res/managed-services/registration-definition`
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes; or I'm proposing the module to be migrated from CARML/TFVM.

### Bicep or Terraform?

Bicep

### Module Classification?

Resource Module

### Module Name

avm/res/managed-services/registration-definition

### Module Details

CARML to AVM

### Do you want to be the owner of this module?

No

### Module Owner's GitHub Username (handle)

_No response_

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by segraef**
## Description

Closes https://github.com/Azure/Azure-Verified-Modules/issues/558

## Pipeline Reference

| Pipeline |
| -------- |
|     [avm.res.managed-services.registration-definition](https://github.com/segraef/bicep-registry-modules/actions/workflows/avm.res.managed-services.registration-definition.yml)     |

## Type of Change

- [x] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: .github/CODEOWNERS
Comment (segraef) [inline .github/CODEOWNERS:76]: pending approval

**Comment by segraef** [comment]
Unsure why [this](https://github.com/segraef/bicep-registry-modules/actions/runs/8567995514/job/23480942345) fails but not locally?

image

File: .github/workflows/avm.res.network.network-watcher.yml
Comment (AlexanderSehr) [inline .github/workflows/avm.res.network.network-watcher.yml:1]: Seams like this snuck in here ;)

File: avm/res/managed-services/registration-definition/main.bicep
Comment (AlexanderSehr) [inline avm/res/managed-services/registration-definition/main.bicep:70]: If I'm not mistaken, we talked about designing an approach that flattens this module, right? So that both the parent and the child could be published by making each a top-level module.

Comment (AlexanderSehr) [reply avm/res/managed-services/registration-definition/main.bicep:70]: Just checked and saw we wanted to have a call about this like ... mid March :D

File: avm/res/managed-services/registration-definition/registration-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/res/managed-services/registration-definition/registration-assignment/main.bicep:11]: As this is a different resource and not a 'child-module'  it should be placed in a file & folder `managed-services/registration-definition/modules/registrationAssignment.bicep`.

That also eliminates the need for a dedicated readme & version.json file.

**Comment by AlexanderSehr** [comment]
> Unsure why [this](https://github.com/segraef/bicep-registry-modules/actions/runs/8567995514/job/23480942345) fails but not locally?
> 
> image

Hey @segraef,
if that worked locally, we have a problem 😄 

The log says 
`Expected $null or empty, because each parameter in the template file should have a description starting with a "Category" prefix like "Required. " and ending with a dot. Found incorrect items: [registrationAssignmentId, registrationDefinitionId]., but got @('registrationAssignmentId', 'registrationDefinitionId').` 

So it finds 2 parameters that are missing the category. Looking at the file, both are indeed missing `Required.` in front of the parameter.
image

File: avm/res/managed-services/registration-definition/registration-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/res/managed-services/registration-definition/registration-assignment/main.bicep:9]:
```suggestion
@description('Required. The ID of the registration definition.')
param registrationDefinitionId string

@description('Required. The ID of the registration assignment.')
param registrationAssignmentId string
```

File: .github/workflows/avm.res.network.network-watcher.yml
Comment (segraef) [reply .github/workflows/avm.res.network.network-watcher.yml:1]: I don't know where this is coming from 😶‍🌫️

**Comment by segraef** [comment]
> > Unsure why [this](https://github.com/segraef/bicep-registry-modules/actions/runs/8567995514/job/23480942345) fails but not locally?
> > image
> 
> Hey @segraef, if that worked locally, we have a problem 😄
> 
> The log says `Expected $null or empty, because each parameter in the template file should have a description starting with a "Category" prefix like "Required. " and ending with a dot. Found incorrect items: [registrationAssignmentId, registrationDefinitionId]., but got @('registrationAssignmentId', 'registrationDefinitionId').`
> 
> So it finds 2 parameters that are missing the category. Looking at the file, both are indeed missing `Required.` in front of the parameter. image

Oh thank you so much, for some reason it was not obvious to me.

File: avm/res/managed-services/registration-definition/registration-assignment/main.bicep
Comment (segraef) [reply avm/res/managed-services/registration-definition/registration-assignment/main.bicep]: Adjusted.

**Comment by segraef** [comment]
Lighthouse configuration required, give me some time, thanks.

**Comment by AlexanderSehr** [comment]
> Lighthouse configuration required, give me some time, thanks.

Sure @segraef. @matebarabas & @jtracey93 please also advice which secondary tenant to point this to. The one of the CARML subscription would come to mind...

File: avm/res/managed-services/registration-definition/tests/e2e/max/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/managed-services/registration-definition/tests/e2e/max/main.test.bicep:1]: Truth be told, I tihnk we should remove this test case as it's not required and doesn't test more than the waf-test

**Comment by matebarabas** [comment]
> > Lighthouse configuration required, give me some time, thanks.
> 
> Sure @segraef. @matebarabas & @jtracey93 please also advice which secondary tenant to point this to. The one of the CARML subscription would come to mind...

@AlexanderSehr, I checked with @segraef and he mentioned that for initial testing, he can use one of his subs. The question to me is how we are going to cover our periodic tests. I'd defer to @jtracey93 and/or would ask how did we end up solving the same problem for the AAD module?

**Comment by AlexanderSehr** [comment]
> > > Lighthouse configuration required, give me some time, thanks.
> > 
> > 
> > Sure @segraef. @matebarabas & @jtracey93 please also advice which secondary tenant to point this to. The one of the CARML subscription would come to mind...
> 
> @AlexanderSehr, I checked with @segraef and he mentioned that for initial testing, he can use one of his subs. The question to me is how we are going to cover our periodic tests. I'd defer to @jtracey93 and/or would ask how did we end up solving the same problem for the AAD module?

Hey @matebarabas, if you say 'initial testing' do you refer to 'in his fork' or upstream in the 'brm repository'. Only the latter will enable the module to be published as without the test would just fail and the module not be published. I guess this is known, yet I wanted to mention it just in case.

The same is to be said about the periodic tests. They will only work / not work if we have a valid secondary tenant to test with.

Regarding AADDS we're good. It did not need a 2nd tenant but only additional permissions to be able to deploy the service.

**Comment by segraef** [comment]
Manual deployment works after having groups created to 

image


Template deployment worked but getting random avmTelemetry errors and other random InternalServerErrors after trying additional deployments which I can't explain.

image

**Comment by segraef** [comment]
Seems to work now: [avm.res.managed-services.registration-definition](https://github.com/segraef/bicep-registry-modules/actions/workflows/avm.res.managed-services.registration-definition.yml).
Need to fix the `avmTelemetry` stuff now, so wait for me.

**Comment by segraef** [comment]
Green now.
[avm.res.managed-services.registration-definition](https://github.com/segraef/bicep-registry-modules/actions/workflows/avm.res.managed-services.registration-definition.yml)

File: avm/res/managed-services/registration-definition/main.bicep
Comment (AlexanderSehr) [inline avm/res/managed-services/registration-definition/main.bicep:27]: Would not recommend to call it `resourceLocation` if it's not a resourceLocation but either something like 
- location
- metadataLocation
- telemetryLocation

Comment (segraef) [reply avm/res/managed-services/registration-definition/main.bicep:27]: Wait, so it's used everywhere as `resourceLocation` but here I shouldn't?

**Comment by segraef** [comment]
> > Lighthouse configuration required, give me some time, thanks.
> 
> Sure @segraef. @matebarabas & @jtracey93 please also advice which secondary tenant to point this to. The one of the CARML subscription would come to mind...

used the CARML tenant by the way

**Comment by segraef** [comment]
> Green now. [avm.res.managed-services.registration-definition](https://github.com/segraef/bicep-registry-modules/actions/workflows/avm.res.managed-services.registration-definition.yml)

still shows red but is green, red light green light ...

File: avm/res/managed-services/registration-definition/main.bicep
Comment (AlexanderSehr) [reply avm/res/managed-services/registration-definition/main.bicep:27]: Hey @segraef it 'should' be used everywhere as a resource location. Like the deployment of a key vault which is deployed into a location. In this case it's only used for the telemetry deployment object, or isn't it?

**Comment by AlexanderSehr** [comment]
> > Green now. [avm.res.managed-services.registration-definition](https://github.com/segraef/bicep-registry-modules/actions/workflows/avm.res.managed-services.registration-definition.yml)
> 
> still shows red but is green, red light green light ...

Can take a moment 😄 . Looks good now

File: avm/res/managed-services/registration-definition/tests/e2e/max/main.test.bicep
Comment (AlexanderSehr) [reply avm/res/managed-services/registration-definition/tests/e2e/max/main.test.bicep:1]: cc: @segraef

File: avm/res/managed-services/registration-definition/main.bicep
Comment (segraef) [reply avm/res/managed-services/registration-definition/main.bicep]: @AlexanderSehr, changed.

File: avm/res/managed-services/registration-definition/tests/e2e/max/main.test.bicep
Comment (AlexanderSehr) [reply avm/res/managed-services/registration-definition/tests/e2e/max/main.test.bicep:1]: ok, I guess you did not see this comment, nor the reminder. I still stand by it and think we should definitely remove the test, but I'll resolve the comment to not block the PR over it.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**
